source/extensions/filters/http/jwt_authn/authenticator.cc

```diff
@@ -66,87 +80,94 @@ class AuthenticatorImpl : public Logger::Loggable<Logger::Id::filter>,
   // The HTTP request headers
   Http::HeaderMap* headers_{};
   // The on_done function.
-  Authenticator::Callbacks* callback_{};
+  AuthenticatorCallback callback_;
 
   // The pending uri_, only used for logging.
   std::string uri_;
   // The pending remote request so it can be canceled.
   Http::AsyncClient::Request* request_{};
+  // Check audience object for overriding the providers.
+  ::google::jwt_verify::CheckAudiencePtr audiences_;
+  // specific issuer or not.
+  const absl::optional<std::string> issuer_;
+  const bool is_allow_failed_;
 };
 
-void AuthenticatorImpl::sanitizePayloadHeaders(Http::HeaderMap& headers) const {
-  for (const auto& it : config_->getProtoConfig().providers()) {
-    const auto& provider = it.second;
-    if (!provider.forward_payload_header().empty()) {
-      headers.remove(Http::LowerCaseString(provider.forward_payload_header()));
-    }
-  }
-}
-
-void AuthenticatorImpl::verify(Http::HeaderMap& headers, Authenticator::Callbacks* callback) {
+void AuthenticatorImpl::verify(Http::HeaderMap& headers, std::vector<JwtLocationConstPtr>&& tokens,
+                               AuthenticatorCallback callback) {
   headers_ = &headers;
-  callback_ = callback;
+  tokens_ = std::move(tokens);
+  callback_ = std::move(callback);
 
   ENVOY_LOG(debug, "Jwt authentication starts");
-  auto tokens = config_->getExtractor().extract(headers);
-  if (tokens.empty()) {
-    if (okToBypass()) {
-      doneWithStatus(Status::Ok);
-    } else {
-      doneWithStatus(Status::JwtMissed);
-    }
+  if (tokens_.empty()) {
+    doneWithStatus(Status::JwtMissed);
     return;
   }
 
-  // TODO(qiwzhang), add supports for multiple tokens.
-  // Only process the first token for now.
-  token_.swap(tokens[0]);
+  startVerify();
+}
 
-  const Status status = jwt_.parseFromString(token_->token());
-  if (status != Status::Ok) {
-    doneWithStatus(status);
-    return;
-  }
+void AuthenticatorImpl::startVerify() {
+  Status status;
```

Comment:
initialize status.

Reply:
status is initialized as OK by default constructor

Reply:
Do you want me to initialize it as something other than OK?

Reply:
@lizan you are right.  I mixed this Status with others.  Yes we need to initialize it.